relayer/provider/cosmos/provider.go

```diff
@@ -888,16 +888,89 @@ func (cc *CosmosProvider) MsgTransfer(amount sdk.Coin, dstChainId, dstAddr, srcP
 // MsgRelayTimeout constructs the MsgTimeout which is to be sent to the sending chain.
 // The counterparty represents the receiving chain where the receipts would have been
 // stored.
-func (cc *CosmosProvider) MsgRelayTimeout(ctx context.Context, dst provider.ChainProvider, dsth int64, packet provider.RelayPacket, dstChanId, dstPortId, srcChanId, srcPortId string) (provider.RelayerMessage, error) {
+func (cc *CosmosProvider) MsgRelayTimeout(
+	ctx context.Context,
+	dst provider.ChainProvider,
+	dsth int64,
+	packet provider.RelayPacket,
+	dstChanId, dstPortId, srcChanId, srcPortId string,
+	order chantypes.Order,
+) (provider.RelayerMessage, error) {
 	var (
 		acc string
 		err error
+		msg provider.RelayerMessage
 	)
 	if acc, err = cc.Address(); err != nil {
 		return nil, err
 	}
 
+	switch order {
+	case chantypes.UNORDERED:
+		msg, err = cc.unorderedChannelTimeoutMsg(ctx, dst, dsth, packet, acc, dstChanId, dstPortId, srcChanId, srcPortId)
+		if err != nil {
+			return nil, err
+		}
+	case chantypes.ORDERED:
+		msg, err = cc.orderedChannelTimeoutMsg(ctx, dst, dsth, packet, acc, dstChanId, dstPortId, srcChanId, srcPortId)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("invalid order type %s, order should be %s or %s",
+			order.String(), chantypes.ORDERED.String(), chantypes.UNORDERED.String())
+	}
+
+	return msg, nil
+}
+
+func (cc *CosmosProvider) orderedChannelTimeoutMsg(
+	ctx context.Context,
+	dst provider.ChainProvider,
+	dsth int64,
+	packet provider.RelayPacket,
+	acc, dstChanId, dstPortId, srcChanId, srcPortId string,
+) (provider.RelayerMessage, error) {
+	seqRes, err := dst.QueryNextSeqRecv(ctx, dsth, dstChanId, dstPortId)
+
+	switch {
+	case err != nil:
+		return nil, err
+	case seqRes.Proof == nil:
+		return nil, fmt.Errorf("timeout packet next sequence received proof seq(%d) is nil", packet.Seq())
+	case seqRes == nil:
+		return nil, fmt.Errorf("timeout packet [%s]seq{%d} has no associated proofs", cc.PCfg.ChainID, packet.Seq())
+	default:
+		msg := &chantypes.MsgTimeout{
+			Packet: chantypes.Packet{
+				Sequence:           packet.Seq(),
+				SourcePort:         srcPortId,
+				SourceChannel:      srcChanId,
+				DestinationPort:    dstPortId,
+				DestinationChannel: dstChanId,
+				Data:               packet.Data(),
+				TimeoutHeight:      packet.Timeout(),
+				TimeoutTimestamp:   packet.TimeoutStamp(),
+			},
+			ProofUnreceived:  seqRes.Proof,
+			ProofHeight:      seqRes.ProofHeight,
+			NextSequenceRecv: packet.Seq(),
+			Signer:           acc,
+		}
+
+		return NewCosmosMessage(msg), nil
+	}
+}
+
+func (cc *CosmosProvider) unorderedChannelTimeoutMsg(
+	ctx context.Context,
+	dst provider.ChainProvider,
+	dsth int64,
+	packet provider.RelayPacket,
+	acc, dstChanId, dstPortId, srcChanId, srcPortId string,
+) (provider.RelayerMessage, error) {
 	recvRes, err := dst.QueryPacketReceipt(ctx, dsth, dstChanId, dstPortId, packet.Seq())
+
```

Comment:
Same issue on the switch after this line (but commented here because it's part of the diff).